### DISCUSSION TITLE
Advanced tags. Autocue. LS 2.3.x.

### DISCUSSION THREAD

**Discussion by gAlleb** [AzuraCast/AzuraCast#7560] [General]
Hi there!

I suspect that when `azuracast.autocue` is disabled and you are using another autocue implementation or NOT using any implementation at all — LS won't understand custom `Advanced tags` settings per track basis. Is this the case?

As I see from a code — tags are now designed specifically for `azuracast.autocue`. If they are present — autocue will fetch them, if not - it will calculate them and cache them. But it won't work if we disable autocue or use another implementation. I might be wrong here but that's what I'm seeing.

Say someone has disabled autocue and wants to set custom `cue_in` value. Will it work? Or will be presented just like `azuracast_autocue=\"{\\\"cue_in\\\":50.0...` metadata and will be ignored with disabled autocue?

**Comment by unudelaradio** [comment]
@gAlleb I think you are right! I've just tested the latest Rolling Release. I have to mention that i'm a **Visual Custom Cue Points** fan user. I'm not using AutoCue and i prefer the native AzuraCast method: `Normal Mode` with `2 seconds` Crossfade Duration.

1. So in this native method case some custom cues are ignored: `fade in` is the strongest example (i've set it to `0` and is totally ignored). Custom Start Next (`overlap`) is also ignored, but only when there's no `fade out` set. Custom `cue in` and `cue out` are respected.

2. But since i don't like `fade in` for my songs, i disabled crossfading and paste this code in the 3rd LS Config box:
```
settings.azuracast.default_fade := 0.00
settings.azuracast.default_cross := 0.15
settings.azuracast.enable_crossfade := true
settings.azuracast.crossfade_type := "normal"
```
Here's what i get:
Custom `fade in`: still not respected, but since `default_fade` is now `0.00` i don't mind, because all my custom `fade in` tags were already `0` for all my songs.
Custom `cue in`, `cue out`, `overlap` and `fade out` works fine.

**Comment by RM-FM** [comment]
@gAlleb As I understand it, the AzuraCast Autocue integration is always active, regardless of whether you have enabled autocue in the profile settings or not - it simply doesn't run `autocue.internal` if it is not enabled. Therefore, it appears to be designed to function even without the autocue setting being enabled. However, if you overrule/exclude the AzuraCast Autocue provider, you are likely correct that manually set cue points and fades are ignored.

@unudelaradio I believe this is a bug that should be reported as an issue. From what I understand, manual cue points and fades should also function correctly in manual/normal mode. It’s possible that the issue lies in handling zero values, as there have been similar problems in the past where a value of "0" was mistakenly treated as "no value."

**Reply by gAlleb** [reply]
Yeah, I suppose you are right. It's designed in a way that first it checks `azuracast_autocue` regardless of whether `settings.azuracast.compute_autocue` is true or false

```
def azuracast.autocue(
    ~request_metadata,
    ~file_metadata,
    filename
) =
    label="azuracast.autocue"

    # Check if `"azuracast_autocue"` is present; If so, use azuracast autocue implementation:
    if list.assoc.mem("azuracast_autocue", request_metadata) then

```

But  cue_file unregisters this. 

Hm, maybe there is a way to keep it...

**Comment by BusterNeece** [comment]
I've reverted some changes we made to our AutoCue data so that we now once again send regular annotations for our DB-stored AutoCue metadata. This means that we're now sending the amplify data in as `liq_amplify` so that it'll set regardless of AutoCue's enabled status, and the other annotations are named things like `autocue_fade_in`, `autocue_fade_out`, etc. so it can more easily be used by third-party implementations.

**Reply by gAlleb** [reply]
Thanks, Buster.

```
2024/11/23 22:23:56 [azuracast.api:3] API nextsong - Response (200): {"uri":"annotate:title=\"Crystal Ball\",artist=\"I See Stars\",duration=\"223.43\",song_id=\"33c81235a2aaa76295d45bb7999834a6\",media_id=\"2792\",sq_id=\"180684\",azuracast_autocue=\"true\",liq_amplify=\"-12.06 dB\",autocue_cue_in=\"0.50\",autocue_cue_out=\"222.70\",autocue_fade_in=\"0.10\",autocue_fade_out=\"1.10\",playlist_id=\"3\":media:rock/I See Stars/New Demons/10 Crystal Ball.mp3"}
```

And this one `azuracast_autocue=\"true\"`? Could you clarify?
I've switched `autocue` off in settings to use cue_file. Here it says it's true.

**Reply by BusterNeece** [reply]
@gAlleb That's just an annotation that tells our own LS-side implementation that we have DB-based AutoCue data and should use that instead of recomputing it. If you're doing your own implementation, out of the box it won't do anything, but it can be used as a reference if you want to use the DB-side data.

**Reply by gAlleb** [reply]
Yeah, got it. I got it when I hit the `comment button` :) 
So theoretically I can make some changes to cue_file to read and understand this values - then convert them to its `liq_` format.

But as I understand you've decided not to populate DB with autocue results when it calculates the values? Autocue computes -> values saved -> next time files is already annotated with those values. Hm. Probably to grant user a privilege of controlling manually every value and set it as he likes. I see. Yeah. But on the other hand if track doesn't have manual values - autocue (when enabled) caches them for 1 session. Am I getting it right?

**Reply by BusterNeece** [reply]
@gAlleb Yep, Moonbase made a pretty strong case that it would be a nightmare for us to cache, at the DB level, AutoCue calculations that happen during one particular run of Liquidsoap, because of the myriad of variables that can change from run to run. It makes much more sense for us to just cache those values in a runtime SQLite database that Liquidsoap itself can hit, resulting in far better performance when the cache is hit and leaving AzuraCast completely out of the picture.

**Reply by gAlleb** [reply]
I see. Thanks!
